Add debug visualiser for 'dict' and implement IReadOnlyDictionary/IReadOnlyCollection

saul opened the pull request:
Fixes #3981, cc @0x53A 

![image](https://user-images.githubusercontent.com/64654/33043576-d50ceb38-ce3d-11e7-80c9-d91af06a73a2.png)

saul commented:
@cartermp this can be merged

dhwed commented:
While you're at it, can you also implement `IReadOnlyDictionary<'Key,'T>`?

cartermp commented:
@saul @0x53A does this work with the following program?

```fsharp
[<EntryPoint>]
let main argv = 
    let d = dict [("uh",1);("uhhh",2);("uhhhhhhhhhh",3)]
    let d2 = dict [("hello", 15); ("Wow", 10)]
    (d, d2) |> ignore // breakpoint here
    0 // return an integer exit code
fsharp
```

Perhaps it's environmental, but when I install this PR into VS 15.5 Preview 4 I still see the old behavior.

saul commented:
Can’t test it at the moment - my screenshot is from manually linking the new FSharp.Core into my project.

0x53A commented:
@cartermp I've downloaded FSharp.Core.dll from https://ci2.dot.net/job/Microsoft_visualfsharp/job/master/job/release_net40_no_vs_windows_nt_prtest/4613/artifact/release/net40/bin/

And it looks really good:
![grafik](https://user-images.githubusercontent.com/4236651/33137883-6179fc34-cfa9-11e7-97b5-f44a278cbff3.png)


Since this is part of FSharp.Core, not the VF#Tools, this would need a new nuget release.

cartermp commented:
I did a `git clean -xdfsmorgasburg`, rebuilt things, and it all works. 👍

cartermp commented:
@dhwed For `IReadOnlyDictionary`, I've created a tracking issue: #3999.  I'd rather that be done separately from this PR, but if @saul is open to quickly doing it or accepting a PR into his branch from you, it won't be the end of the world if it comes in through this PR.

forki commented on src/fsharp/FSharp.Core/fslib-extra-pervasives.fs:
```diff
@@ -40,59 +40,72 @@ module ExtraTopLevelOperators =
+        [<DebuggerBrowsable(DebuggerBrowsableState.RootHidden)>]
+        member x.Items = Array.ofSeq d
+
+    let dictImpl (comparer:IEqualityComparer<'SafeKey>) (makeSafeKey : 'Key->'SafeKey) (getKey : 'SafeKey->'Key) (l:seq<'Key*'T>) : IDictionary<'Key,'T> =
```
No longer inlined.

forki replied:
?

saul replied:
Good catch, fixed

saul commented:
Implemented IReadOnlyDictionary and IReadOnlyCollection - I'll go around adding them to list, Map and Set now.

cartermp commented:
https://github.com/fsharp/fslang-design/pull/234

dsyme commented on src/fsharp/FSharp.Core/fslib-extra-pervasives.fs:
```diff
@@ -78,6 +78,13 @@ module ExtraTopLevelOperators =
+        interface IReadOnlyDictionary<'Key, 'T> with
+            member __.Item with get key = t.[makeSafeKey key]
+            member __.Keys = t.Keys |> Seq.map getKey
+            member __.TryGetValue(key, value) = t.TryGetValue(makeSafeKey key, ref value)
```
``ref value`` is incorrect here - please use ``&value``.  Please also add a test for this and review all testing as there must be a test hole here.

saul replied:
Ref seems to work fine for me? I've added tests to validate it - I'll use the `&` instead though.

dsyme replied:
It is my understanding that ``ref`` will not copy-out the resulting value to the byref passed as an argument

saul commented:
@dsyme there's a bit of a hiccup - `dict` currently returns `IDictionary` - this doesn't implement `IReadOnly*` interfaces. We can either create a new function called something like `readOnlyDict` that returns an IReadOnlyDictionary, or we can change `dict` to return some type in FSharp.Core that implements `IDictionary` and `IReadOnly*`, but I think that would break backwards compat.

How do you want to proceed?

dsyme commented:
> @dsyme there's a bit of a hiccup - dict currently returns IDictionary - this doesn't implement IReadOnly* interfaces. We can either create a new function called something like readOnlyDict that returns an IReadOnlyDictionary, or we can change dict to return some type in FSharp.Core that implements IDictionary and IReadOnly*, but I think that would break backwards compat.

My assumption was that for now the return type should be only ``IDictionary`` and that would allow the change to be integrated straight away.  

If you want to add ``readOnlyDict`` then please either add that to the RFC or make a new RFC and lets discuss there?  I've no real issue with adding it but new API should always have multiple eyes and an RFC

saul commented:
Makes sense - will update the RFC tonight and use this PR as its implementation.

KevinRansom commented:
@dsyme, @saul saul ... is this ready?  It looks like you agreed to a change, but I don't see the change.

Thanks

Kevin

saul commented:
@KevinRansom it's not quite ready yet - I started writing tests and fell down the rabbit hole of adding support for IEnumerable.Reset on `seq {}` enumerators. I'll see if I can make this change without that 😄

saul commented:
Pushed tests and new `readOnlyDict` top-level operator. Discussion to continue in https://github.com/fsharp/fslang-suggestions/issues/624 and then the RFC will need updating, too.

saul commented:
Hmm not sure why the CI is failing... I've added that file to source control? cc @KevinRansom

cloudRoutine commented:
@saul what do you think of creating an additional project just for debugger visualizers? (not necessarily in this PR, but something to consider for the future)
The vsix packaging would [install the dll of the visualizers](https://docs.microsoft.com/en-us/visualstudio/debugger/how-to-install-a-visualizer) to the right directory and vs should auto-load it
Although I'm not sure if these visualizers would show up in C# as well, but if we only write custom visualizers for F# types it'll probably be fine.
- [writing a visualizer C#](https://docs.microsoft.com/en-us/visualstudio/debugger/walkthrough-writing-a-visualizer-in-csharp)
- [visualizer api](https://docs.microsoft.com/en-us/visualstudio/debugger/visualizer-api-reference)

saul commented:
Visualizer in that sense is completely different to debug type proxies that are used in FSharp.Core.

dsyme commented:
Yes, I still believe this line is buggy: https://github.com/Microsoft/visualfsharp/pull/3988/files#diff-cf1d05f204f61f31020312fcec812dbaR84.  That ``ref`` is reading the provided byref argument and creating a new ref cell, and never writing the result back to the byref argument.

saul commented:
Yeah @dsyme I’m happy to change it, but the TryGetValue tests pass (if you try it locally, the CI is failing)

0x53A commented on src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/ExtraTopLevelOperatorsTests.fs:
```diff
@@ -0,0 +1,213 @@
+        Assert.AreEqual(irod.Values, [| 1; 4; 9|])
+        
+        Assert.IsTrue(irod.TryGetValue(1, ref 1))
+        Assert.IsFalse(irod.TryGetValue(100, ref 1))
```
As I read it, the test only verifies that the key is found, __not__ that the correct value is returned. That may explain why it passes even though @dsyme is sure that the ref is incorrect.

A better test may be

```F#
let b,i = irod.TryGetValue(2)
Assert.AreEqual(true, b)
Assert.AreEqual(2, i)
```

cartermp commented:
(For posterity, since the jenkins servers to wipe results after a time)
```
11:42:56 Errors and Failures
11:42:56 
11:42:56 1) Failed : FSharp.Core.UnitTests.FSharp_Core.Microsoft_FSharp_Core.DictTests.IReadOnlyDictionary on readOnlyDict
11:42:56   Expected: 4
11:42:56   But was:  0
11:42:56 at FSharp.Core.UnitTests.FSharp_Core.Microsoft_FSharp_Core.DictTests.IReadOnlyDictionary on readOnlyDict() in D:\j\w\release_ci_pa---866fd2c3\tests\FSharp.Core.UnitTests\FSharp.Core\Microsoft.FSharp.Core\ExtraTopLevelOperatorsTests.fs:line 116
11:42:56 
11:42:56 2) Failed : FSharp.Core.UnitTests.SurfaceArea.SurfaceAreaTest.VerifyArea
11:42:56 Assembly: FSharp.Core, Version=4.4.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
11:42:56 Expected and actual surface area don't match. To see the delta, run:
11:42:56     windiff D:\j\w\release_ci_pa---866fd2c3\tests\FSharp.Core.UnitTests\SurfaceArea.net40.fs D:\j\w\release_ci_pa---866fd2c3\release\net40\bin\CoreUnit_net40_Xml.xml
11:42:56 
11:42:56 Unexpectedly missing (expected, not actual):
11:42:56 
11:42:56 Unexpectedly present (actual, not expected):
11:42:56     Microsoft.FSharp.Core.ExtraTopLevelOperators: System.Collections.Generic.IReadOnlyDictionary`2[TKey,TValue] CreateReadOnlyDictionary[TKey,TValue](System.Collections.Generic.IEnumerable`1[System.Tuple`2[TKey,TValue]])
11:42:56 at FSharp.Core.UnitTests.LibraryTestFx.SurfaceArea.verify(String expected, String platform, String fileName) in D:\j\w\release_ci_pa---866fd2c3\tests\FSharp.Core.UnitTests\LibraryTestFx.fs:line 142
11:42:56 
11:42:56 Run Settings
11:42:56     RuntimeFramework: V4.0
11:42:56     WorkDirectory: D:\j\w\release_ci_pa---866fd2c3\release\net40\bin
11:42:56     Verbose: True
11:42:56     NumberOfTestWorkers: 4
11:42:56 
11:42:56 Test Run Summary
11:42:56     Overall result: Failed
11:42:56    Tests run: 5344, Passed: 5342, Errors: 0, Failures: 2, Inconclusive: 0
11:42:56      Not run: 3, Invalid: 0, Ignored: 0, Explicit: 3, Skipped: 0
11:42:56   Start time: 2018-02-02 19:36:10Z
11:42:56     End time: 2018-02-02 19:41:41Z
11:42:56     Duration: 330.699 seconds
11:42:56 
11:42:56 Results (nunit3) saved as D:\j\w\release_ci_pa---866fd2c3\tests\TestResults\test-net40-coreunit-results.xml
```

saul commented:
At least it builds now :) the problem here is that seq computation expressions / Seq.map etc don’t support IEnumerable.Reset(). I just need to refactor the tests keeping this in mind

saul commented:
Scratch that, no it’s not. It’s the TryGetValue issue dsyme raised and the surface area change.

saul commented:
Okay, I think I've fixed the broken tests here. Sorry for the delay - my laptop makes it impossible to do dev work and I've just fixed my workstation after 3 months of it being broken 😄

saul commented:
@dotnet-bot test Ubuntu14.04 Release_default Build please

saul commented:
@dotnet-bot test Ubuntu14.04 Release_default please

cartermp commented:
> Sorry for the delay 

No worries at all!

cartermp commented:
@dsyme can you re-review? Thanks!

dsyme commented:
LGTM

cartermp commented:
@dotnet-bot test Windows_NT Release_ci_part3 Build please

KevinRansom commented:
The issue is this:

an unexpected new public API:
Microsoft.FSharp.Core.ExtraTopLevelOperators: **System.Collections.Generic.IReadOnlyDictionary`2[TKey,TValue] CreateReadOnlyDictionary[TKey,TValue](System.Collections.Generic.IEnumerable`1[System.Tuple`2[TKey,TValue]])**

````
1) Failed : FSharp.Core.UnitTests.Portable.SurfaceArea.SurfaceAreaTest.VerifyArea
Assembly: FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Expected and actual surface area don't match. To see the delta, run:
    windiff D:\j\w\release_ci_pa---c6929ad7\tests\FSharp.Core.UnitTests\SurfaceArea.coreclr.fs \My Documents\CoreUnit_coreclr_Xml.xml

Unexpectedly missing (expected, not actual):

Unexpectedly present (actual, not expected):
    Microsoft.FSharp.Core.ExtraTopLevelOperators: System.Collections.Generic.IReadOnlyDictionary`2[TKey,TValue] CreateReadOnlyDictionary[TKey,TValue](System.Collections.Generic.IEnumerable`1[System.Tuple`2[TKey,TValue]])
at FSharp.Core.UnitTests.LibraryTestFx.SurfaceArea.verify(String expected, String platform, String fileName) in D:\j\w\release_ci_pa---c6929ad7\tests\FSharp.Core.UnitTests\LibraryTestFx.fs:line 171

Run Settings
    Internal Trace: Off

Test Run Summary
  Overall result: Failed
  Test Count: 5389, Passed: 5384, Failed: 1, Inconclusive: 0, Skipped: 4
    Failed Tests - Failures: 1, Errors: 0, Invalid: 0
    Skipped Tests - Ignored: 2, Explicit: 2, Other: 0
  Start time: 2018-03-14 17:52:10Z
    End time: 2018-03-14 17:56:18Z
    Duration: 248.537 seconds


````

KevinRansom commented:
@dotnet-bot test this please

saul commented:
@KevinRansom @cartermp this is ready to merge

KevinRansom commented:
@saul 

Thanks for this.

Kevin
